### PR TITLE
feat: unflatten event before outputting

### LINKED ISSUE #354: Restore event structure into original structure before encoding
### Description

We need to ensure that nested JSON is exploded back into it's original
nested form before being written downstream.

### Example

```json
{
  "parent.child": "value",
  "array[1]": "value1",
  "array[2]": "value2"
}
```

Shoud be exploded into it's nested form before being written downstream:

```json
{
  "parent": {
    "child": "value"
  },
  "array": ["value1", "value2"]
}
```

### Solution

This needs to be a low-level default encoding behavior that occurs before
we encode an event, regardless of the purpose or sink. The way we represent
nested keys internally should be hidden from downstream observers.

### Requirements

- [x] This should be the case for _all_ sinks that encode events.
- [x] This should handle nested keys (`.` delimited keys)
- [x] This should handle indexed keys (`key[1]`, `key[2]`, etc)

### PR DESCRIPTION
**Pull request by LucioFranco**
This PR provides the ability to unflatten structured events as described in #354. This takes a somewhat complicated approach by building each tree path individually then merging them as we go. Performance seems to actually be not that bad. Showing only a 2x slow down for deeply nested items.

Closes #354

### CODE REVIEW

**Comment by binarylogic** [comment]
> Showing only a 2x slow down for deeply nested items.

Do you have specific benchmarks for this? I'm curious what your definition of "deeply nested" is here. Are we talking 3 or more levels?

**Comment by LucioFranco** [comment]
@binarylogic right roughly what I expect is 3 being somewhat the max of what we might usually see. It may be worth it to do some more thorough benchmarks.

File: src/event/unflatten.rs [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+        /// The strategy will then recursively create random enum structures
+        /// using the leaf strategy that only creates `bool`, `i64` and `[a-z]+`
+        /// strings.
+        ///
+        /// It will then use a strategy of creating either a `vec` or a `hash_map`.
+        pub fn json() -> impl Strategy<Value = Value> {
+            let leaf = prop_oneof![
+                any::<bool>().prop_map(Value::Bool),
+                any::<i64>().prop_map(|n| Value::Number(n.into())),
+                "[a-z]+".prop_map(Value::String),
```
Comment (lukesteensen) [inline src/event/unflatten.rs]: This is the part that worries me. If we want to merge this, we should have a really clear idea of the different patterns that will trip it up (e.g. actual JSON keys like `"foo.bar"`) and think about how common we expect those to be.

Comment (LucioFranco) [reply src/event/unflatten.rs]: I think possibly what we can also do is provide some form of escaping for things like `.` and `[\d+]`.

cc @binarylogic

**Comment by binarylogic** [comment]
> This is super complicated, so I'm basically leaning on the proptest stuff to be comfortable that it works.

I dislike this. And agree, this is very complicated. First, it's hard to trust that this will actually work and perform properly. Second, this is going to be very hard to maintain and debug. I think it's work creating a spec for this and getting consensus on a better direction.

>  If we're good with that caveat, I'm fine to merge this.

I think we need to merge this because it is such as fundamental issue right now. It's blocking other sink implementations, but I think we should address this ASAP with a follow up PR.

> But it might be worth thinking about how much more effort it would be to properly support nested object in our internal representation and sidestep this problem entirely.

Agree.

---

The address the `.` reserved character issue. This is documented in our [data model](https://docs.vector.dev/about/data-model) page. If a user sends a key with `.` it will be exploded into a nested object. This is pretty common across many other services from what I've seen, so I'm ok with this caveat. If we do want to address it my thoughts are:

1. Escape it, as @LucioFranco pointed out. Ex: `parent\\.child`
2. Actually adjust our internal representation of this data so it is stored in a nested format and does not require a special delimiter.

**Pull request merged**